woocommerce/woocommerce-gutenberg-products-block
# Lazy Loading Atomic Components

This PR introduces the use of Lazy Loading for the Atomic Blocks (aka Product Elements) Components used on the frontend of the store.

The main benefit of this change is that the main package sizes are reduced due to not needing to include the atomic blocks. Atomic blocks have their own packages which are imported dynamically only when needed. As our catalog of atomic blocks increase, this change will be more beneficial.

It works as follows:

1. The file `components-init.js` is loaded when the Components are needed (prior to rendering a product layout).
2. This file defines the dynamic webpack public path based on the plugin directory (`__webpack_public_path__ = WC_BLOCKS_BUILD_URL;`)
3. This file then registers each Block Component with a lazy callback. Example:

```jsx
registerBlockComponent( {
	blockName: 'woocommerce/product-rating',
	component: lazy( () =>
		import(
			/* webpackChunkName: "atomic-block-components/rating" */ './product/rating/block'
		)
	),
} );
```

When files are built, Webpack sees the chunk names and creates a .js file that will be imported by the client. In the above case, this file will be found in `build/atomic-block-components/rating.js`

The component that facilitates rendering on these blocks wraps them with a `Suspense` block which shows a loading state until they are ready.

**To test:**

Smoke test both the All Products Block and Single Products Block. Both of these are using Atomic Blocks and both implement `Suspense`. Ensure you check all atomic blocks are rendered correctly on the frontend.

**Changelog**

> Implemented Lazy Loading for Product Blocks to improve performance.
